internal/pkg/controller/controller.go

```diff
@@ -0,0 +1,332 @@
+		switch event.eventType {
+			case "create":
+				ObjectCreated(obj, c.client)
+			case "update":
+				ObjectUpdated(obj, c.client)
+			case "delete":
+				ObjectDeleted(obj)
+		}
+	}
+	return nil
+}
+
+// handleErr checks if an error happened and makes sure we will retry later.
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+		c.queue.Forget(key)
+		return
+	}
+
+	// This controller retries 5 times if something goes wrong. After that, it stops trying.
+	if c.queue.NumRequeues(key) < 5 {
+		logrus.Infof("Error syncing events %v: %v", key, err)
+
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	runtime.HandleError(err)
+	logrus.Infof("Dropping the key %q out of the queue: %v", key, err)
+}
+
+// ObjectCreated Do nothing for default handler
+func ObjectCreated(obj interface{}, client kubernetes.Interface) {
+	message := "Configmap: `" + obj.(*v1.ConfigMap).Name + "`has been created in Namespace: `" + obj.(*v1.ConfigMap).Namespace + "`"
+	logrus.Infof(message)
+	err := rollingUpgradeDeployments(obj, client)
+	if err != nil {
+		logrus.Errorf("failed to update Deployment: %v", err)
+	}
+}
+
+// ObjectDeleted Do nothing for default handler
+func ObjectDeleted(obj interface{}) {
+
+}
+
+// ObjectUpdated Do nothing for default handler
+func ObjectUpdated(oldObj interface{}, client kubernetes.Interface) {
+	message := "Configmap: `" + oldObj.(*v1.ConfigMap).Name + "`has been updated in Namespace: `" + oldObj.(*v1.ConfigMap).Namespace + "`"
+	logrus.Infof(message)
+	err := rollingUpgradeDeployments(oldObj, client)
+	if err != nil {
+		logrus.Errorf("failed to update Deployment: %v", err)
+	}
+}
+
+// Implementation has been borrowed from fabric8io/configmapcontroller
+// Method has been modified a little to use updated liberaries.
+func rollingUpgradeDeployments(oldObj interface{}, client kubernetes.Interface) error {
+	ns := oldObj.(*v1.ConfigMap).Namespace
+	configMapName := oldObj.(*v1.ConfigMap).Name
+	configMapVersion := convertConfigMapToToken(oldObj.(*v1.ConfigMap))
+
+	deployments, err := client.Apps().Deployments(ns).List(meta_v1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list deployments")
+	}
+	for _, d := range deployments.Items {
+		containers := d.Spec.Template.Spec.Containers
+		// match deployments with the correct annotation
+		annotationValue, _ := d.ObjectMeta.Annotations[updateOnChangeAnnotation]
```

Comment:
should write annotationValue := d.ObjectMeta.Annotations[updateOnChangeAnnotation] instead of annotationValue, _ := d.ObjectMeta.Annotations[updateOnChangeAnnotation]